floris/simulation/flow_field.py

```diff
@@ -127,11 +127,15 @@ def initialize_velocity_field(self, grid: Grid) -> None:
         # for height, using it here to apply the shear law makes that dimension store the vertical
         # wind profile.
         wind_profile_plane = (grid.z_sorted / self.reference_wind_height) ** self.wind_shear
-        dwind_profile_plane = (
-            self.wind_shear
-            * (1 / self.reference_wind_height) ** self.wind_shear
-            * (grid.z_sorted) ** (self.wind_shear - 1)
-        )
+        if self.wind_shear == 0.0:
+            # Avoid division by zero at z = 0.0 when wind_shear = 0.0
```

Comment:
It occurs when `self.wind_shear` is zero and `grid.z_sorted` happens to include the value zero (a sample point at the ground level). For instance, removing the `ìf` statement and modifing example 29 by replacing
```python
profiles = get_profiles('y') + get_profiles('z')
````
with
```python
profiles_y = get_profiles('y')
profile_range_z = [-hub_height, 3 * D]
profiles_z = fi.sample_velocity_deficit_profiles(
	direction='z',
	downstream_dists=downstream_dists,
	profile_range=profile_range_z,
	homogeneous_wind_speed=homogeneous_wind_speed,
)
profiles = profiles_y + profiles_z

````
gives the following warnings
```
/home/uname/python/floris/floris/simulation/flow_field.py:137: RuntimeWarning: divide by zero encountered in reciprocal
  * (grid.z_sorted) ** (self.wind_shear - 1)
/home/uname/python/floris/floris/simulation/flow_field.py:135: RuntimeWarning: invalid value encountered in multiply
  self.wind_shear
````
This is because the first element in `profile_range_z` makes the method include a sample point at ground level.

Reply:
Thanks for clarifying! I see now: the $z^{(\alpha-1)}$ evaluates to $1/0$ when $z=0$ and $\alpha=0$. This produces an `inf` and raises a divide by zero warning; then, the multiplication by $\alpha=0$ raises an invalid multiplication warning (0 x `inf`) and results in a `nan` in `dwind_profile_plane` (the derivative of the shear profile layer).

However, I actually think we should let these warnings be raised. In my testing, they only appear once, and they let the user know that something strange is happening when one evaluates the derivative of the shear profile at $z=0$ (the derivative is not well defined). In fact the divide by zero problem occurs when $z=0$ for any value of shear less than 1 (not just when shear is 0). The resulting figures appear to be the same regardless.

@rafmudaf @bayc , what are our practices regarding letting warnings be raised in "unusual" circumstances?

Reply:
Ok @vallbog @misi9170 I'm just coming up from my obsession on this question for the past couple of weeks. I started by trying to create a small unit test for this situation and realized that the Grid classes weren't able to be created from the `from_dict` method because they relied on the `Vec3` class, so that was addressed in #751. In the process of that, I noticed that many of the classes in `floris.simulation` weren't proper slotted classes because the base classes weren't attrs-derived, and this exposed quite a bit of monkey patching - all of this was fixed in #750. Finally, I've been able to come back to this and now we can better test this particular situation through the unit testing infrastructure. I've written a general issue for this in #760.

In testing this further, I noticed that the `if wind_shear == 0.0` condition is inadequate since the issue comes from the exponent in the derivate being less than 1 which causes the "divide by zero in reciprocal" runtime warning. As @misi9170 pointed out, this will happen if wind_shear is less than 1. However, it's actually the combination of wind shear less than 1 _and_ z=0 that is problematic. To accommodate that, I propose to initialize the derivate to zeros and use `np.power` to calculate the derivative since it is a ufunc and supports the `where=` option:

```python
        dwind_profile_plane = (
            self.wind_shear
            * (1 / self.reference_wind_height) ** self.wind_shear
            * np.power(
                grid.z_sorted,
                (self.wind_shear - 1),
                where=grid.z_sorted != 0.0
            )
        )
```

I initially used `np.where`, but that evaluates all operations and then picks out the ones that fit the condition so the runtime warning is still raised.

Here's a very simple script to test:
```python
fi = FlorisInterface("inputs/gch.yaml")
x = [0.0] #, 0.0]
y = [0.0] #, 0.0]
z = [0.0] #, 10.0]
u_at_points = fi.sample_flow_at_points(x, y, z)
```

Without the proposed change, this raises a divide by zero warning.

Reply:
@rafmudaf Thanks, this all makes sense. Feel free to merge. That `np.where()` evaluating both and therefore still raising the issue is something I've run into before, and is a pretty annoying limitation. But that's another story...